File: cpp/kskipmrr.cpp
#include <string>

#include "util/MyTypes.h"
#include "util/MyBlas.h"
#include "util/MyNpy.h"
#include "util/MyTimer.h"

#include "cnpy.h" // npy

using namespace std;
using namespace MyTypes;
using namespace MyNpy;
using namespace MyBlas;
using namespace cnpy;

vector<double> kskipmrr(const Matrix &A, const vector<double> &b, const double epsilon, const int k);

int main() {
  string matrixFilePath = "../data/matrix.npy";
  string vectorFilePath = "../data/vector.npy";

  Matrix A = loadMatrix(matrixFilePath);
  vector<double> b = loadVector(vectorFilePath);
  const double epsilon = 1e-8;
  const unsigned int k = 0; 

  // timer
  std::chrono::system_clock::time_point  start, end;
  start = std::chrono::system_clock::now();
  auto x = kskipmrr(A, b, epsilon, k);
  end = std::chrono::system_clock::now();
  double elapsed = chrono::duration_cast<std::chrono::milliseconds>(end-start).count();
  cout << elapsed << "ms"<< endl;
  printVec(x);
  return 0;
}

vector<double> kskipmrr(const Matrix &A, const vector<double> &b, const double epsilon, const int k){
  const ul N = A.size();
  const ul max_iter = N*2;
  const double b_norm = vecvec(b, b);
	vector<double> x(N, 0.0);

  // 初期化
  vv Ar(k + 2, vector<double>(N));
  vv Ay(k + 1, vector<double>(N));
  vector<double> alpha(2*k+3);
  vector<double> beta(2*k+2);
  vector<double> delta(2*k+1);
  beta[0] = 0;

  // 初期残差
  Ar[0] = b - matvec(A, x);
  double residual = vecvec(Ar[0], Ar[0]) / b_norm;

  // 初期反復
  Ar[1] = matvec(A, Ar[0]);
  double zeta = vecvec(Ar[0], Ar[1]) / vecvec(Ar[1], Ar[1]);
  Ay[0] = zeta * Ar[1];
  vector<double> z = -zeta * Ar[0];
  Ar[0] = Ar[0] - Ay[0];
  x = x - z;
  ui i = 1;
  ui index = 1;

  // 反復計算
  while (i < max_iter) {
    // 収束判定
    residual = vecvec(Ar[0], Ar[0]) / b_norm;
    if (residual < epsilon) break;

    // 事前計算
    int jj;
    for (int j=1;j<k+2;j++) Ar[j]=matvec(A, Ar[j-1]);
    for (int j=1;j<k+1;j++) Ay[j]=matvec(A, Ay[j-1]);
    for (int j=0;j<2*k+3;j++) { 
      jj = j;
      alpha[j] = vecvec(Ar[jj], Ar[jj + j % 2]);
    }
    for (int j=1;j<2*k+2;j++) { 
      jj = j;
      beta[j] = vecvec(Ay[jj], Ar[jj + j % 2]);
    }
    for (int j=0;j<2*k+1;j++) { 
      jj = j;
      delta[j] = vecvec(Ay[jj], Ay[jj + j % 2]);
    }

    // MrRでの1反復(解の更新)
    double d = alpha[2] * delta[0] - pow(beta[1],2);
    zeta = alpha[1] * delta[0] / d;
    double eta = -alpha[1] * beta[1] / d;
    Ay[0] = eta * Ay[0] + zeta * Ar[1];
    z = eta * z - zeta * Ar[0];
    Ar[0] = Ar[0] - Ay[0];
    Ar[1] = matvec(A, Ar[0]);
    x = x - z;

    // MrRでのk反復
    for (int j=0;j<k;j++) {
      delta[0] = pow(zeta,2) * alpha[2] + eta * zeta * beta[1];
      alpha[0] = alpha[0] - zeta * alpha[1];
      delta[1] = pow(eta,2) * delta[1] + 2 * eta * zeta * beta[2] + pow(zeta,2) * alpha[3];
      beta[1] = eta * beta[1] + zeta * alpha[2] - delta[1];
      alpha[1] = -beta[1];
      for (int l=2;l<2*(k-j)+1;l++) {
        delta[l] = pow(eta,2) * delta[l] + 2 * eta * zeta * beta[l+1] + pow(zeta,2) * alpha[l + 2];
        double tau = eta * beta[l] + zeta * alpha[l + 1];
        beta[l] = tau - delta[l];
        alpha[l] = alpha[l] - tau + beta[l];
      }
      // 解の更新
      d = alpha[2] * delta[0] - pow(beta[1],2);
      zeta = alpha[1] * delta[0] / d;
      eta = -alpha[1] * beta[1] / d;
      Ay[0] = eta * Ay[0] + zeta * Ar[1];
      z = eta * z - zeta * Ar[0];
      Ar[0] = Ar[0] - Ay[0];
      Ar[1] = matvec(A, Ar[0]);
      x = x - z;
    }

    i += (k + 1);
    index++;
  }
  
  return x;
}

File: v2/gpu/kskipmrr.py
import cupy as cp
from cupy import dot
from cupy.linalg import norm

from .common import start, finish, init, MultiGpu


def kskipmrr(A, b, epsilon, k, T):
    # 初期化
    MultiGpu.init_gpu(0, 1)
    A, b, x, b_norm, N, max_iter, residual, num_of_solution_updates = init(A, b, T, 2)
    MultiGpu.alloc(A, b, T)

    # 初期化
    Ar = cp.empty((k + 2, N), T)
    Ay = cp.empty((k + 1, N), T)
    alpha = cp.empty(2 * k + 3, T)
    beta = cp.empty(2 * k + 2, T)
    delta = cp.empty(2 * k + 1, T)
    beta[0] = 0

    # 初期残差
    Ar[0] = b - MultiGpu.dot(A, x)
    residual[0] = norm(Ar[0]) / b_norm

    # 初期反復
    start_time = start(method_name='k-skip MrR', k=k)
    Ar[1] = MultiGpu.dot(A, Ar[0])
    zeta = dot(Ar[0], Ar[1]) / dot(Ar[1], Ar[1])
    Ay[0] = zeta * Ar[1]
    z = -zeta * Ar[0]
    Ar[0] -= Ay[0]
    x -= z
    num_of_solution_updates[1] = 1
    i = 1
    index = 1

    # 反復計算
    while i < max_iter:
        # 収束判定
        residual[index] = norm(Ar[0]) / b_norm
        if residual[index] < epsilon:
            isConverged = True
            break

        # 基底計算
        for j in range(1, k + 2):
            Ar[j] = MultiGpu.dot(A, Ar[j-1])
        for j in range(1, k + 1):
            Ay[j] = MultiGpu.dot(A, Ay[j-1])

        # 係数計算
        for j in range(2 * k + 3):
            jj = j // 2
            alpha[j] = dot(Ar[jj], Ar[jj + j % 2])
        for j in range(1, 2 * k + 2):
            jj = j//2
            beta[j] = dot(Ay[jj], Ar[jj + j % 2])
        for j in range(2 * k + 1):
            jj = j // 2
            delta[j] = dot(Ay[jj], Ay[jj + j % 2])

        # MrRでの1反復(解の更新)
        d = alpha[2] * delta[0] - beta[1] ** 2
        zeta = alpha[1] * delta[0] / d
        eta = -alpha[1] * beta[1] / d
        Ay[0] = eta * Ay[0] + zeta * Ar[1]
        z = eta * z - zeta * Ar[0]
        Ar[0] -= Ay[0]
        Ar[1] = MultiGpu.dot(A, Ar[0])
        x -= z

        # MrRでのk反復
        for j in range(k):
            delta[0] = zeta ** 2 * alpha[2] + eta * zeta * beta[1]
            alpha[0] -= zeta * alpha[1]
            delta[1] = eta ** 2 * delta[1] + 2 * eta * \
                zeta * beta[2] + zeta ** 2 * alpha[3]
            beta[1] = eta * beta[1] + zeta * alpha[2] - delta[1]
            alpha[1] = -beta[1]
            for l in range(2, 2 * (k - j) + 1):
                delta[l] = eta ** 2 * delta[l] + 2 * eta * \
                    zeta * beta[l+1] + zeta ** 2 * alpha[l + 2]
                tau = eta * beta[l] + zeta * alpha[l + 1]
                beta[l] = tau - delta[l]
                alpha[l] -= tau + beta[l]
            # 解の更新
            d = alpha[2] * delta[0] - beta[1] ** 2
            zeta = alpha[1] * delta[0] / d
            eta = -alpha[1] * beta[1] / d
            Ay[0] = eta * Ay[0] + zeta * Ar[1]
            z = eta * z - zeta * Ar[0]
            Ar[0] -= Ay[0]
            Ar[1] = MultiGpu.dot(A, Ar[0])
            x -= z

        i += (k + 1)
        index += 1
        num_of_solution_updates[index] = i
    else:
        isConverged = False
        residual[index] = norm(Ar[0]) / b_norm

    elapsed_time = finish(start_time, isConverged, i, residual[index])
    return elapsed_time, num_of_solution_updates[:index+1], residual[:index+1]
